astropy/units/quantity.py

```diff
@@ -372,123 +356,39 @@ def __array_prepare__(self, obj, context=None):
-                    try:
-                        converters[0] = units[0]._get_converter(
-                            dimensionless_unscaled)
-                    except UnitConversionError:
-                        raise exc
-                    else:
-                        result_unit = dimensionless_unscaled
-
+        # determine required converter functions -- to bring the unit of the
+        # input to that expected (e.g., radian for np.sin), or to get
+        # consistent units between two inputs (e.g., in np.add) --
+        # and the unit of the result
+        converters, result_unit = converters_and_unit(function, '__call__',
+                                                      *args)
+        if function.nout > 1:
+            result_unit = result_unit[context[2]]
         # We now prepare the output object
         if self is obj:
-
             # this happens if the output object is self, which happens
             # for in-place operations such as q1 += q2
 
-            # In some cases, the result of a ufunc should be a plain Numpy
-            # array, which we can't do if we are doing an in-place operation.
-            if result_unit is None:
-                raise TypeError("Cannot store non-quantity output from {0} "
-                                "function in {1} instance"
-                                .format(function.__name__, type(self)))
-
-            if self.__quantity_subclass__(result_unit)[0] is not type(self):
-                raise UnitTypeError(
-                    "Cannot store output with unit '{0}' from {1} function "
-                    "in {2} instance.  Use {3} instance instead."
-                    .format(result_unit, function.__name__, type(self),
-                            self.__quantity_subclass__(result_unit)[0]))
-
-            # If the Quantity has an integer dtype, in-place operations are
-            # dangerous because in some cases the quantity will be e.g.
-            # decomposed, which involves being scaled by a float, but since
-            # the array is an integer the output then gets converted to an int
-            # and truncated.
-            result_dtype = np.result_type(*((args + (float,))
-                                            if any(converters) else args))
-            if not np.can_cast(result_dtype, obj.dtype, casting='same_kind'):
-                raise TypeError("Arguments cannot be cast safely to inplace "
-                                "output with dtype={0}".format(self.dtype))
-
-            result = self  # no view needed since we return the object itself
+            # Check that we're not trying to store a plain Numpy array or a
+            # Quantity with an inconsistent unit (e.g., not angular for Angle),
+            # and that we can handle the type (e.g., that we are not int when
+            # float is required).
+            check_output(obj, result_unit, (args + tuple(
```

Comment:
`check_output` will internally get a common dtype for all those args. Now suppose I have two integer quantities `q1` and `q2` (these are difficult to construct, but still allowed...), with units `km` and `m` and try `q1 += q2` -- in this case, the converter will just be a division by 1000, so the if test fails, and the `float16` ensures that the `check_output` will test that the output argument can handle `float` (it is `float16` as the minimum length float, just in case `q1` and `q2` are also `float16`).

I must admit the case for `int8` is less clear -- I think I just added it as a small integer alternative.

To be honest, looking at this again, I'm also not sure it isn't better to just go with with what was above (except replace `float` with `float16` since there is no reason to disallow that). Though, then again, I think I even added a test for the above but with `q1` in `m` and `q2` in `km`, in which case integers are OK...